compiler/rustc_trait_selection/src/solve/project_goals.rs

```diff
@@ -551,6 +551,26 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
             .to_predicate(tcx),
         )
     }
+
+    fn consider_builtin_discriminant_kind_candidate(
+        ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+    ) -> QueryResult<'tcx> {
+        let self_ty = goal.predicate.self_ty();
+
+        let tcx = ecx.tcx();
+        let term = self_ty.discriminant_ty(tcx).into();
+
+        Self::consider_assumption(
+            ecx,
+            goal,
+            ty::Binder::dummy(ty::ProjectionPredicate {
+                projection_ty: tcx.mk_alias_ty(goal.predicate.def_id(), [self_ty]),
+                term,
+            })
+            .to_predicate(tcx),
+        )
```

Comment:
I'd prefer just to equate here -- makes it clearer what it's doing, and the trait has no substs other than the self type. This code can then get even shorter once `eq_term_and_make_canonical_response` lands from #107348.

```suggestion
        let discriminant = goal.predicate.self_ty().discriminant_ty(ecx.tcx());
        let nested_goals = ecx.infcx.eq(goal.predicate.term.expect_ty(), discriminant);
        ecx.evaluate_all_and_make_canonical_response(nested_goals)
```